### ISSUE TITLE
NotNullIfNotNull attribute does not work with binary operators

### ISSUE THREAD

**Issue by duracellko** [dotnet/roslyn#48489]
**Version Used**: 
.NET 5 SDK RC1

**Steps to Reproduce**:

Define following operator:

```csharp
[return: NotNullIfNotNull("x")]
[return: NotNullIfNotNull("y")]
public static Key? operator +(Key? x, Key? y)
{
    ...
}
```

And use it as following:

```csharp
Key result = new Key('a') + new Key('b')
```

**Expected Behavior**:

Compiler does not report any warning.

**Actual Behavior**:

Compiler reports warning: `warning CS8604: Possible null reference`

**Details**

I was able to reproduce it in unit test:

```csharp
        [Fact]
        public void NotNullIfNotNull_Return_BinaryOperator()
        {
            var lib_cs =
@"using System.Diagnostics.CodeAnalysis;
public class C
{
    [return: NotNullIfNotNull(""x""), NotNullIfNotNull(""y"")] public static C? operator +(C? x, C? y) => null;
    public static C? operator *(C? x, C? y) => null;
    [return: NotNullIfNotNull(""x"")] public static C? operator -(C? x, C? y) => null;
    [return: NotNullIfNotNull(""y"")] public static C? operator /(C? x, C? y) => null;
}";
            var source = @"
class D
{
    static void M(C c1, C c2, C? cn1, C? cn2)
    {
        (c1 + c2).ToString(); // no warn
        (c1 + cn2).ToString(); // no warn
        (cn1 + c2).ToString(); // no warn
        (cn1 + cn2).ToString(); // warn

        (c1 * c2).ToString(); // warn
        (c1 * cn2).ToString(); // warn
        (cn1 * c2).ToString(); // warn
        (cn1 * cn2).ToString(); // warn

        (c1 - c2).ToString(); // no warn
        (c1 - cn2).ToString(); // no warn
        (cn1 - c2).ToString(); // warn
        (cn1 - cn2).ToString(); // warn
        
        (c1 / c2).ToString(); // no warn
        (c1 / cn2).ToString(); // warn
        (cn1 / c2).ToString(); // no warn
        (cn1 / cn2).ToString(); // warn
    }
}";
            var expected = new[]
            {
                // (9,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 + cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 + cn2").WithLocation(9, 10),
                // (11,10): warning CS8602: Dereference of a possibly null reference.
                //         (c1 * c2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1 * c2").WithLocation(11, 10),
                // (12,10): warning CS8602: Dereference of a possibly null reference.
                //         (c1 * cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1 * cn2").WithLocation(12, 10),
                // (13,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 * c2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 * c2").WithLocation(13, 10),
                // (14,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 * cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 * cn2").WithLocation(14, 10),
                // (18,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 - c2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 - c2").WithLocation(18, 10),
                // (19,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 - cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 - cn2").WithLocation(19, 10),
                // (22,10): warning CS8602: Dereference of a possibly null reference.
                //         (c1 / cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1 / cn2").WithLocation(22, 10),
                // (24,10): warning CS8602: Dereference of a possibly null reference.
                //         (cn1 / cn2).ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "cn1 / cn2").WithLocation(24, 10),
            };

            var lib = CreateNullableCompilation(new[] { NotNullIfNotNullAttributeDefinition, lib_cs });
            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
            comp.VerifyDiagnostics(expected);

            var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullIfNotNullAttributeDefinition });
            comp2.VerifyDiagnostics(expected);
            CompileAndVerify(comp2);
        }
```

Probably same problem is with `NotNullWhen` attribute.

**Comment by duracellko** [comment]
This issue may be similar to closed issue #39802
